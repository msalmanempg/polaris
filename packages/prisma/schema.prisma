//******** AUTO GENERATED FILE, DO NOT EDIT.  *********

datasource db {
    provider = "postgresql"
    url      = env("DATABASE_URL")
}

generator client {
    provider = "prisma-client-js"
}

generator dbml {
    provider = "prisma-dbml-generator"
}


model Agreement {
  id                   Int                 @id @default(autoincrement())
  leadId               Int                 @map("lead_id")
  agreementNumber      String              @unique @map("agreement_number")
  publishedPrice       Int                 @map("published_price")
  bookingPrice         Int                 @map("booking_price")
  discountValue        Int?                @map("discount_value")
  posessionValue       Int                 @map("posession_value")
  payments             ExpectedPayment[]
  customers            AgreementCustomer[]
  salesPerson          SalePerson[]
  nominees             Nominee[]
  Unit                 Unit                @relation(fields: [unitId], references: [id])
  unitId               Int                 @map("unit_id")
  status               agreementStatus
  bookingDate          DateTime            @map("booking_date")
  notarisedDate        DateTime            @map("notarised_date")
  agreementDate        DateTime            @map("agreement_date")
  installmentStartDate DateTime            @map("installment_start_date")
  installmentEndDate   DateTime            @map("installment_end_date")
  createdAt            DateTime            @default(now()) @map("created_at")
  updatedAt            DateTime            @updatedAt @map("updated_at")

  @@map("agreements")
}

model DraftAgreement {
  id              Int      @id @default(autoincrement())
  Unit            Unit     @relation(fields: [unitId], references: [id])
  unitId          Int      @map("unit_id")
  agreementNumber String   @unique @map("agreement_number")
  data            Json     @map("data")
  createdAt       DateTime @default(now()) @map("created_at")
  updatedAt       DateTime @updatedAt @map("updated_at")

  @@map("draft_agreements")
}

model SalePerson {
  id          Int       @id @default(autoincrement())
  designation String
  fullName    String    @map("full_name")
  email       String
  govtId      String    @map("govt_id")
  phoneNumber String    @map("phone_number")
  sortLevel   Int       @map("sort_level")
  Agreement   Agreement @relation(fields: [agreementId], references: [id])
  agreementId Int       @map("agreement_id")
  createdAt   DateTime  @default(now()) @map("created_at")
  updatedAt   DateTime  @updatedAt @map("updated_at")

  @@map("sale_persons")
}

model AgreementCustomer {
  id                   Int                @id @default(autoincrement())
  type                 customerType       @map("customer_type")
  Customer             Customer           @relation(fields: [customerId], references: [id])
  customerId           Int                @map("customer_id")
  Agreement            Agreement          @relation(fields: [agreementId], references: [id])
  agreementId          Int                @map("agreement_id")
  isPrimary            Boolean            @default(true)
  meta                 Json?
  email                String
  primaryPhoneNumber   String             @map("primary_phone_number")
  secondaryPhoneNumber String?            @map("secondary_phone_number")
  address              String
  Companies            AgreementCompany[]
  createdAt            DateTime           @default(now()) @map("created_at")
  updatedAt            DateTime           @updatedAt @map("updated_at")

  @@map("agreement_customers")
}

model AgreementCompany {
  id                  Int               @id @default(autoincrement())
  AgreementCustomer   AgreementCustomer @relation(fields: [agreementCustomerId], references: [id])
  agreementCustomerId Int               @map("agreement_customer_id")
  name                String
  ntn                 String
  registrationNumber  String            @unique @map("registration_number")
  address             String
  city                String
  email               String
  contactNumber       String            @map("contact_number")
  meta                Json?
  createdAt           DateTime          @default(now()) @map("created_at")
  updatedAt           DateTime          @updatedAt @map("updated_at")

  @@map("agreement_companies")
}

model Nominee {
  id            Int        @id @default(autoincrement())
  govtId        String     @map("govt_id")
  govtIdType    govtIdType @map("govt_id_card")
  nomineeFor    String     @map("nominee_for")
  fullName      String     @map("full_name")
  contactNumber String     @map("contact_number")
  email         String
  relationship  String
  Agreement     Agreement  @relation(fields: [agreementId], references: [id])
  agreementId   Int        @map("agreement_id")
  meta          Json?
  createdAt     DateTime   @default(now()) @map("created_at")
  updatedAt     DateTime   @updatedAt @map("updated_at")

  @@map("nominees")
}

model AgreementAttachment {
  id           Int                   @id @default(autoincrement())
  agreementId  Int                   @map("agreement_id")
  Attachment   Attachment            @relation(fields: [attachmentId], references: [id])
  attachmentId Int
  type         agreementDocumentType
  createdAt    DateTime              @default(now()) @map("created_at")
  updatedAt    DateTime              @updatedAt @map("updated_at")

  @@map("agreement_attachments")
}

model NomineeAttachment {
  id           Int                   @id @default(autoincrement())
  nomineeId    Int                   @map("nominee_id")
  Attachment   Attachment            @relation(fields: [attachmentId], references: [id])
  attachmentId Int
  type         nomineeAttachmentType
  createdAt    DateTime              @default(now()) @map("created_at")
  updatedAt    DateTime              @updatedAt @map("updated_at")

  @@map("nominee_attachment")
}

enum govtIdType {
  cnic
  nicop
  poc
}

enum customerType {
  individual
  company
}

enum agreementStatus {
  draft
  review
  active
  transferred
  terminated
}

enum agreementDocumentType {
  agreement
  booking_form
  payment_plan
  floor_plan
  other
}

enum nomineeAttachmentType {
  id_front
  id_back
  profile_picture
  other
}


model Attachment {
  id                   Int                   @id @default(autoincrement())
  name                 String
  key                  String
  customerAttachments  CustomerAttachment[]
  nomineeAttachments   NomineeAttachment[]
  agreementAttachments AgreementAttachment[]
  createdAt            DateTime              @default(now()) @map("created_at")
  updatedAt            DateTime              @updatedAt @map("updated_at")

  @@map("attachments")
}


model Customer {
    id             Int                  @id @default(autoincrement())
    fullName       String               @map("full_name")
    govtId         String               @map("govt_id")
    govtIdType     govtIdType           @map("govt_id_type")
    relationType   relationType         @map("relation_type")
    guardianName   String               @map("guardian_name")
    nationality    String
    passportNumber String?              @map("passport_number")
    dateOfBirth    DateTime             @map("date_of_birth")
    gender         Gender
    meta           Json?
    attachments    CustomerAttachment[]
    agreements     AgreementCustomer[]
    createdAt      DateTime             @default(now()) @map("created_at")
    updatedAt      DateTime             @updatedAt @map("updated_at")

    @@unique([govtId, govtIdType], name: "Customer_govtId_govtIdType_unique_constraint")
    @@map(name: "customers")
}

model CustomerAttachment {
    id           Int        @id @default(autoincrement())
    Attachment   Attachment @relation(fields: [attachmentId], references: [id])
    attachmentId Int
    Customer     Customer   @relation(fields: [customerId], references: [id])
    customerId   Int
    createdAt    DateTime   @default(now()) @map("created_at")
    updatedAt    DateTime   @updatedAt @map("updated_at")

    @@map("customer_attachments")
}

enum relationType {
    father
    husband
}

enum Gender {
    male
    female
}

enum customerDocumentType {
    id_front
    id_back
    profile_picture
    other
}


model ExpectedPayment {
  id          Int                   @id @default(autoincrement())
  amount      Int
  percentage  Int?
  paymentDate DateTime              @map("payment_date")
  type        expectedPaymentType
  Agreement   Agreement             @relation(fields: [agreementId], references: [id])
  agreementId Int                   @map("agreement_id")
  meta        Json?
  dueDate     DateTime              @map("due_date")
  status      expectedPaymentStatus
  createdAt   DateTime              @default(now()) @map("created_at")
  updatedAt   DateTime              @updatedAt @map("updated_at")

  @@map("expected_payments")
}

enum expectedPaymentType {
  down_payment
  installment
  advance
  posession
}

enum expectedPaymentStatus {
  paid
  pending
}


model Project {
  id             Int                 @id @default(autoincrement())
  name           String
  address        String
  city           String
  province       String
  country        String
  longitude      Float
  latitude       Float
  email          String
  phone          String
  url            String
  completionDate DateTime?           @map("completion_date")
  logo           String
  meta           Json?
  bankDetails    ProjectBankDetail[]
  units          Unit[]
  createdAt      DateTime            @default(now()) @map("created_at")
  updatedAt      DateTime            @updatedAt @map("updated_at")

  @@map("projects")
}

model ProjectBankDetail {
  id            Int      @id @default(autoincrement())
  bankName      String   @map("bank_name")
  accountTitle  String   @map("account_title")
  accountNumber String   @map("account_number")
  iban          String
  swiftCode     String   @map("swift_code")
  branchAddress String   @map("branch_address")
  branchCode    String   @map("branch_code")
  meta          Json?
  Project       Project? @relation(fields: [projectId], references: [id])
  projectId     Int?
  createdAt     DateTime @default(now()) @map("created_at")
  updatedAt     DateTime @updatedAt @map("updated_at")

  @@map("project_bank_details")
}


model Unit {
  id             Int              @id @default(autoincrement())
  unitNumber     String           @unique @map("unit_number")
  location       String
  type           unitType
  bed            Int
  basePrice      Int              @map("base_price")
  publishedPrice Int              @map("published_price")
  netArea        Int              @map("net_area")
  grossArea      Int              @map("gross_area")
  meta           Json?
  status         unitStatus
  Project        Project          @relation(fields: [projectId], references: [id])
  projectId      Int              @map("project_id")
  agreements     Agreement[]
  drafts         DraftAgreement[]
  completionDate DateTime?        @map("completion_date")
  createdAt      DateTime         @default(now()) @map("created_at")
  updatedAt      DateTime         @updatedAt @map("updated_at")

  @@map(name: "units")
}

enum unitType {
  residential
  commercial
}

enum unitStatus {
  booked
  sold
  available
}
