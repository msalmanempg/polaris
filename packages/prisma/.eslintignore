**/__snapshots__/
**/.ng_build/
**/build/
**/coverage/
**/dist/
**/lib/
**/node_modules/
jest.config.js
*.d.ts