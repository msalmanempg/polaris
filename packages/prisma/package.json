{
  "name": "@polaris/prisma",
  "version": "1.0.0",
  "description": "data layer for Polaris",
  "main": "lib/index.js",
  "repository": "https://github.com/SectorLabs/polaris/tree/main/packages/prisma",
  "license": "MIT",
  "scripts": {
    "build": "yarn ttsc -p tsconfig.json",
    "lint:errors": "eslint --quiet --ext .js,.ts ./",
    "lint:types": "ttsc --noEmit"
  }
}
