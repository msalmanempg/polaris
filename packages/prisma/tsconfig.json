{
  "compilerOptions": {
    "allowJs": true,
    "module": "commonjs",
    "esModuleInterop": true,
    "moduleResolution": "node",
    "skipLibCheck": true,
    "isolatedModules": true,
    "forceConsistentCasingInFileNames": true,
    "lib": ["es2017", "dom"],
    "target": "es2017",
    "types": ["node"],
    "strict": true,
    "baseUrl": ".",
    "paths": {
      "prisma/*": ["."]
    },
    "outDir": "./lib/",
    "sourceMap": true,
    "inlineSources": true,
    "sourceRoot": "/",
    "plugins": [{ "transform": "@zerollup/ts-transform-paths" }]
  },
}
