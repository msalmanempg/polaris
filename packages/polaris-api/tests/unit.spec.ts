import { prisma, Project, Unit } from "@polaris/prisma";
import path from "path";
import { projects, units } from "./support/fixtures/units";
import { testExpressApp } from "./support/requestFactory";

describe("units", () => {
  let createdProject: Project;
  let createdUnits: Unit[];

  beforeEach(async () => {
    const [project] = projects;

    createdProject = await prisma.project.create({
      data: {
        ...(project as any),
      },
    });

    createdUnits = await Promise.all(
      units.map((unit: any) => {
        return prisma.unit.upsert({
          where: { unitNumber: unit.unitNumber },
          update: { ...unit, id: undefined, projectId: createdProject.id },
          create: { ...unit, id: undefined, projectId: createdProject.id },
        });
      })
    );
  });

  afterEach(async () => {
    await prisma.unit.deleteMany({
      where: {
        projectId: {
          in: createdProject.id,
        },
      },
    });

    await prisma.project.deleteMany({
      where: {
        id: {
          equals: createdProject.id,
        },
      },
    });
  });

  it("Should list all units", async () => {
    const params = {
      pageOffset: 0,
      pageSize: 10,
    };

    const { body } = await testExpressApp().get("/api/unit/").query(params).expect(200);

    const createdUnitIds = new Set(createdUnits.map((unit) => unit.id));
    const units = body.results
      .filter((unit: any) => createdUnitIds.has(unit.id))
      .sort((a: { unitNumber: string }, b: { unitNumber: string }) =>
        a.unitNumber.localeCompare(b.unitNumber)
      );

    expect(body).toMatchSnapshot({
      count: expect.any(Number),
      results: expect.any(Array),
    });

    matchUnits(units);
  });

  it("Should list and filter units by project id", async () => {
    const params = {
      pageOffset: 0,
      pageSize: 10,
      projectId: createdProject.id,
    };

    const { body } = await testExpressApp().get("/api/unit").query(params).expect(200);

    body.results.map((unit: any) => expect(unit.projectId).toEqual(createdProject.id));
  });

  it("Should list and filter units by status", async () => {
    const params = {
      pageOffset: 0,
      pageSize: 10,
      status: "available",
    };

    const { body } = await testExpressApp().get("/api/unit").query(params).expect(200);

    body.results.map((unit: any) => expect(unit.status).toEqual("available"));
  });

  it("Should list units with stats", async () => {
    const params = {
      pageOffset: 0,
      pageSize: 10,
      includeStats: true,
    };
    const { body } = await testExpressApp().get("/api/unit").query(params).expect(200);

    expect(body.stats).toBeTruthy();
    expect(body.stats).toBeInstanceOf(Array);

    body.stats
      .sort(
        (a: { status: string; count: number }, b: { status: string; count: number }) =>
          a.status < b.status
      )
      .map((unitStat: { status: string; count: number }) =>
        expect(unitStat).toMatchSnapshot(
          {
            count: expect.any(Number),
          },
          `${unitStat.status}`
        )
      );
  });

  it("Should upload units", async () => {
    const { body } = await testExpressApp()
      .post("/api/unit/import")
      .field("projectId", createdProject.id)
      .attach("units", path.join(__dirname, "support/fixtures/sample.xlsx"));
    matchUnits(body);
  });

  it("Should export units of Project", async () => {
    const { body } = await testExpressApp()
      .get(`/api/unit/export/${createdProject.id}`)
      .expect(200);
    expect(body).toMatchSnapshot();
  });
});

const matchUnits = (units: any) => {
  units.map((unit: any) => {
    expect(unit).toMatchSnapshot({
      id: expect.any(Number),
      createdAt: expect.any(String),
      updatedAt: expect.any(String),
      completionDate: expect.any(String),
      projectId: expect.any(Number),
      unitNumber: expect.any(String),
    });
  });
};
