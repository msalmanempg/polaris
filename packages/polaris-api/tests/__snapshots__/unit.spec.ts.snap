// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`units Should export units of Project 1`] = `"https://localhost:3000/testfile"`;

exports[`units Should list all units 1`] = `
Object {
  "count": Any<Number>,
  "pageSize": 10,
  "results": Any<Array>,
}
`;

exports[`units Should list all units 2`] = `
Object {
  "Project": Object {
    "name": "Project 1 for units",
  },
  "basePrice": 5000000,
  "bed": 3,
  "completionDate": Any<String>,
  "createdAt": Any<String>,
  "grossArea": 1250,
  "id": Any<Number>,
  "location": "Block A, Street 5",
  "meta": null,
  "netArea": 1250,
  "projectId": Any<Number>,
  "publishedPrice": 5500000,
  "status": "available",
  "type": "residential",
  "unitNumber": Any<String>,
  "updatedAt": Any<String>,
}
`;

exports[`units Should list all units 3`] = `
Object {
  "Project": Object {
    "name": "Project 1 for units",
  },
  "basePrice": 10000000,
  "bed": 6,
  "completionDate": Any<String>,
  "createdAt": Any<String>,
  "grossArea": 2500,
  "id": Any<Number>,
  "location": "Block B, Street 6",
  "meta": null,
  "netArea": 2500,
  "projectId": Any<Number>,
  "publishedPrice": 10000000,
  "status": "booked",
  "type": "residential",
  "unitNumber": Any<String>,
  "updatedAt": Any<String>,
}
`;

exports[`units Should list units with stats: all 1`] = `
Object {
  "count": Any<Number>,
  "status": "all",
}
`;

exports[`units Should list units with stats: available 1`] = `
Object {
  "count": Any<Number>,
  "status": "available",
}
`;

exports[`units Should list units with stats: booked 1`] = `
Object {
  "count": Any<Number>,
  "status": "booked",
}
`;

exports[`units Should upload units 1`] = `
Object {
  "basePrice": 100000,
  "bed": 1,
  "completionDate": Any<String>,
  "createdAt": Any<String>,
  "grossArea": 230,
  "id": Any<Number>,
  "location": "DHA",
  "meta": null,
  "netArea": 200,
  "projectId": Any<Number>,
  "publishedPrice": 120000,
  "status": "available",
  "type": "residential",
  "unitNumber": Any<String>,
  "updatedAt": Any<String>,
}
`;

exports[`units Should upload units 2`] = `
Object {
  "basePrice": 100000,
  "bed": 1,
  "completionDate": Any<String>,
  "createdAt": Any<String>,
  "grossArea": 230,
  "id": Any<Number>,
  "location": "DHA",
  "meta": null,
  "netArea": 200,
  "projectId": Any<Number>,
  "publishedPrice": 120000,
  "status": "available",
  "type": "residential",
  "unitNumber": Any<String>,
  "updatedAt": Any<String>,
}
`;

exports[`units Should upload units 3`] = `
Object {
  "basePrice": 100000,
  "bed": 1,
  "completionDate": Any<String>,
  "createdAt": Any<String>,
  "grossArea": 230,
  "id": Any<Number>,
  "location": "DHA",
  "meta": null,
  "netArea": 200,
  "projectId": Any<Number>,
  "publishedPrice": 120000,
  "status": "available",
  "type": "residential",
  "unitNumber": Any<String>,
  "updatedAt": Any<String>,
}
`;

exports[`units Should upload units 4`] = `
Object {
  "basePrice": 100000,
  "bed": 1,
  "completionDate": Any<String>,
  "createdAt": Any<String>,
  "grossArea": 230,
  "id": Any<Number>,
  "location": "DHA",
  "meta": null,
  "netArea": 200,
  "projectId": Any<Number>,
  "publishedPrice": 120000,
  "status": "available",
  "type": "residential",
  "unitNumber": Any<String>,
  "updatedAt": Any<String>,
}
`;

exports[`units Should upload units 5`] = `
Object {
  "basePrice": 100000,
  "bed": 1,
  "completionDate": Any<String>,
  "createdAt": Any<String>,
  "grossArea": 230,
  "id": Any<Number>,
  "location": "DHA",
  "meta": null,
  "netArea": 200,
  "projectId": Any<Number>,
  "publishedPrice": 120000,
  "status": "available",
  "type": "residential",
  "unitNumber": Any<String>,
  "updatedAt": Any<String>,
}
`;

exports[`units Should upload units 6`] = `
Object {
  "basePrice": 100000,
  "bed": 1,
  "completionDate": Any<String>,
  "createdAt": Any<String>,
  "grossArea": 230,
  "id": Any<Number>,
  "location": "DHA",
  "meta": null,
  "netArea": 200,
  "projectId": Any<Number>,
  "publishedPrice": 120000,
  "status": "available",
  "type": "residential",
  "unitNumber": Any<String>,
  "updatedAt": Any<String>,
}
`;
